packages/flutter/lib/src/widgets/sliver_coordinator.dart

```diff
@@ -0,0 +1,254 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'framework.dart';
+import 'notification_listener.dart';
+import 'scroll_notification.dart';
+
+/// Function signature for the [SliverCoordinator.callback], which is called each time
+/// its [CustomScrollView] has been laid out.
+///
+/// The second parameter can be used to retrieve layout information when
+/// [CoordinatedSliver.hasLayoutInfo] is true. See [CoordinatedSliver.getSliverConstraints]
+/// and [CoordinatedSliver.getSliverGeometry].
+typedef SliverCoordinatorCallback = void Function(ScrollNotification notification, SliverCoordinatorData data);
+
+class _SliverLayoutInfo {
+  const _SliverLayoutInfo({
+    required this.constraints,
+    required this.geometry
+  });
+
+  final SliverConstraints constraints;
+  final SliverGeometry geometry;
+}
+
+/// Encapsulates a table of layout information for [CoordinatedSliver]s.
+///
+/// An instance of this class is passed to the [SliverCoordinator.callback].
+/// To retreive data from this table use [CoordinatedSliver.hasLayoutInfo]
+/// [CoordinatedSliver.getSliverConstraints], and [CoordinatedSliver.getSliverGeometry].
+class SliverCoordinatorData {
+  final Map<Object, _SliverLayoutInfo> _idToInfo = <Object, _SliverLayoutInfo>{};
+
+  // Add an entry to the layout information table.
+  void _put<T extends _SliverLayoutInfo>(Object id, T info) {
+    _idToInfo[id] = info;
+  }
+
+  // Retreieve an entry from the layout information table.
+  T? _get<T extends _SliverLayoutInfo>(Object id) {
+    return _idToInfo[id] as T?;
+  }
+
+  // Remove an entry from the layout information table.
+  T? _remove<T extends _SliverLayoutInfo>(Object id) {
+    return _idToInfo.remove(id) as T?;
+  }
+
+  // Clear the contents of the layout information table.
+  @protected
+  void _clear() => _idToInfo.clear();
+}
+
+class _SliverCoordinatorScope extends InheritedWidget {
+  const _SliverCoordinatorScope({ required this.data, required super.child });
+
+  final SliverCoordinatorData data;
+
+  @override
+  bool updateShouldNotify(_SliverCoordinatorScope oldWidget) => false;
+}
+
+/// Provides [ScrollNotification]s for its [CustomScrollView] descendant as well as
```

Comment:
This is really interesting, it might be worth expanding on this a bit more.

Slivers don't need scroll notifications, because they know everything about the scroll position and whether or not scrolling is occurring. A good example of this is the floating/snapping sliver persistent header. When scrolling stope, it triggers its own expand/collapse animation:

https://github.com/flutter/flutter/blob/5fb962722a5ab27205ca98a19e3a431b1b8d22aa/packages/flutter/lib/src/rendering/sliver_persistent_header.dart#L615

It actually documents there one of the scenarios this is meant to provide for:
```dart
/// If the header isn't already fully exposed, then scroll it into view.
```

Reply:
Right, the scroll notifications aren't for sliver layout, they're for trigger behaviors that are a consequence of sliver layouts. I can make this clearer.